<a href="https://colab.research.google.com/github/aniket677/Logistic-K-NN-and-Na-ve-Bayes/blob/master/Copy_of_Tanushree_C__Dhongale.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Here: The libraries called pandas and numpy are imported.

In [ ]:
import pandas as pd
import numpy as np

Iris dataset is the data which is being used and it is a dataset in machine learning and statistics. load_iris is a function which is used to load the iris dataset which is included in scikit-learn in the datasets module.

In [ ]:
from sklearn.datasets import load_iris
iris_dataset = load_iris()

Here: the keys of iris dataset are printed which is similar to dictionary and contain keys and values.

In [ ]:
print("Keys of iris_dataset: \n{}".format(iris_dataset.keys()))

Keys of iris_dataset: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


Value of key DESCR is short desciption of the dataset,followings are the points are there which are contains in it:

In [ ]:
val = iris_dataset['DESCR']
start_val = val[:200]
print(start_val + '\n...')

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive
...


Value of key target_names:

In [ ]:
print("Target names: {}".format(iris_dataset['target_names']))

Target names: ['setosa' 'versicolor' 'virginica']


Value of key feature_names:

In [ ]:
print("Features names: \n{}".format(iris_dataset['feature_names']))

Features names: 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


Printing the type of key data:

In [ ]:
print("Type of Data: {}".format(type(iris_dataset['data'])))

Type of Data: <class 'numpy.ndarray'>


The shape of data is printed using shape method. In iris_dataset, there is 150 data points and 4 features.

In [ ]:
print("Shape of Data: {}".format(iris_dataset['data'].shape))

Shape of Data: (150, 4)


Printing the first five samples of the data:

In [ ]:
print("First five columns of data: \n{}".format(iris_dataset['data'][:5]))

First five columns of data: 
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


Printing the type of key target:

In [ ]:
print("Type of target: {}".format(type(iris_dataset['target'])))

Type of target: <class 'numpy.ndarray'>


Printing the shape of target using shape method:

In [ ]:
print("Shape of Target: {}".format(iris_dataset['target'].shape))

Shape of Target: (150,)


Value of key target:

In [ ]:
print("Target: \n{}".format(iris_dataset['target']))

Target: 
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


Here 0 means setosa, 1 means versicolor,and 2 means virginica

## **Training and Testing Data:**

From scikit-learn framework train_test_split function is imported which can be directly use to split the entire iris data into 75-25%.
In scikit-learn, data is denoted with capital X (i.e the length of the sepal, the lenght of the petal, the width the sepal and the width of the petal) and labels by lowercase y (i.e the three categories in which the flower can be divided viz setosa, versicolor, and virginica). This is inspired by standard formulation in mathematics (x) = y.
In the below code, X is used since the data is two dimensional array (a matrix) and y because the target is a one-demensional array(a vector).

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
iris_dataset['data'], iris_dataset['target'], random_state = 0)

The output of the train_test_split function is X_train, X_test, y_train, y_test which are the NumPy arrays. X_train contains 75% and X_test the remaining 25% of the rows of the dataset.

In [ ]:
print("X_train shape: {}".format(X_train.shape))
print("y_train shape: {}".format(y_train.shape))
print("X_test shape: {}".format(X_test.shape))
print("y_test shape: {}".format(y_test.shape))

X_train shape: (112, 4)
y_train shape: (112,)
X_test shape: (38, 4)
y_test shape: (38,)


**This is the K-nearest neighbor machine learning model.**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1)

Training the k-nearest neighbors data:

In [ ]:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

## **Making Predictions**

In [ ]:
X_new = np.array([[5, 2.9, 1, 0.2]])
print("X_new.shape: {}".format(X_new.shape))

X_new.shape: (1, 4)


To make a prediction, we call the predict method of the knn object:

In [ ]:
prediction = knn.predict(X_new)

print("Prediction: {}".format(prediction))
print("Predicted target name: {}".format(
    iris_dataset['target_names'][prediction]
))

Prediction: [0]
Predicted target name: ['setosa']


## **Evaluating the Model**

In [ ]:
y_pred = knn.predict(X_test)
print("Test set predictions: \n{}".format(y_pred))

Test set predictions: 
[2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


Here, the predicted value of the species i.e y_pred is compared with the actual value i.e y_test. If the values are same it gives result as 1 and if not the result is 0 and the mean of all these values are calculated using np.mean() function.

In [ ]:
print("Test set score: {}".format(np.mean(y_pred == y_test)))

Test set score: 0.9736842105263158
